### PR TITLE
ClearML experiment tracking integration

### PR DESCRIPTION
**Pull request by thepycoder**
This PR adds integration with the open-source experiment tracker [ClearML](https://cutt.ly/yolov5-readme-clearml). Installing the package `pip install clearml` will enable the integration and allow users to track every training run in ClearML. This in turn allows users to keep track of different experiments, compare them to see the differences and even run the experiment remotely (using the instructions in the ClearML readme)

### Experiment Tracking and Comparison

After installing and initializing ClearML (with `clearml-init` CLI command), one can run `train.py` with any configuration desired and ClearML integration will automatically be enabled.

ClearML scalars dashboard

What will be captured:
- Source code + uncommitted changes
- Installed packages
- (Hyper)parameters
- Model files (use `--save-period n` to save a checkpoint every n epochs)
- Console output
- Scalars (mAP_0.5, mAP_0.5:0.95, precision, recall, losses, learning rates, ...)
- General info such as machine details, runtime, creation date etc.
- All produced plots such as label correlogram and confusion matrix
- Images with bounding boxes per epoch
- Mosaic per epoch
- Validation images per epoch
- ...

Each of these metrics can easily be compared between multiple experiments using the ClearML web console.

### Versioned Dataset Support

Users will be able to provide a ClearML dataset version as part of the YOLOv5 command line interface, for training. This dataset will be downloaded or taken from cache and used to further train on. 

ClearML Dataset Interface

### Hyperparameter Optimization

A standalone script is provided, which will allow users to run HPO on YOLOv5 locally or in the cloud as well.
HPO

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced YOLOv5 with ClearML integration for advanced ML experiment tracking and management.

### 📊 Key Changes
- Added recommendation for using Weights & Biases Logging.
- Integrated ClearML tool for automatic experiment tracking and dataset versioning.
- Modified README to introduce ClearML integration with its features and links.
- Updated requirements to suggest the installation of ClearML.
- Tweaked `train.py` to support data logging for ClearML.
- Enhanced Jupyter notebook tutorial to include ClearML setup and usage guide.
- Updated `general.py` to recognize 'clearml://' dataset IDs for dataset versioning.
- Introduced ClearML logger in `utils/loggers/__init__.py`.
- Provided a detailed README for ClearML integration at `utils/loggers/clearml/README.md`.
- Crafted a separate ClearML utility module at `utils/loggers/clearml/clearml_utils.py`.
- Included a hypothetical script for hyperparameter optimization (`hpo`) using ClearML at `utils/loggers/clearml/hpo.py`.
- Adjusted `wandb_utils.py` to harmonize with ClearML dataset structures.
- Modified `metrics.py` and `plots.py` to add plot titles for clarity.

### 🎯 Purpose & Impact
- 🚀 **Purpose**: Incorporate ClearML for better experiment tracking, dataset versioning, and model management, offering an alternative to Weights & Biases.
- 🎛 **Impact**:
  - Users gain the ability to trace training runs comprehensively with live updates and detailed metrics.
  - Offers additional capabilities like uncommitted code tracking and reproducibility across machines.
  - Expands dataset management by allowing users to version their datasets using ClearML dataset IDs.
  - Enhances the existing benchmarking system with an exhaustive hyperparameter optimization script.
  - Facilitates easier overview and analysis of experiments through well-organized dashboards and logging systems within the ClearML environment.

### CODE REVIEW

File: utils/loggers/__init__.py [hunk 1]
```diff
@@ -139,9 +166,22 @@ def on_fit_epoch_end(self, vals, epoch, best_fitness, fi):
             with open(file, 'a') as f:
                 f.write(s + ('%20.5g,' * n % tuple([epoch] + vals)).rstrip(',') + '\n')
 
+        # ClearML automatically captures scalars from tensorboard
```
Comment (AyushExel) [inline utils/loggers/__init__.py]: @thepycoder I think some comments won't be necessary. This part is easy to understand 🔥

File: utils/loggers/__init__.py [hunk 2]
```diff
@@ -139,9 +166,22 @@ def on_fit_epoch_end(self, vals, epoch, best_fitness, fi):
...
+        elif self.clearml:
+            for k, v in x.items():
+                title, series = k.split('/')
+                self.clearml.task.get_logger().report_scalar(title, series, v, epoch)
+
+        # Reset epoch image limit
+        if self.clearml:
+            self.clearml.current_epoch_logged_images = set()
+            # ClearML automatically detects epochs using hooks for automatic reporting, but for manual reporting
+            # getting the epoch number from the task itself is slow, so we keep track for internal bookkeeping only
```
Comment (AyushExel) [inline utils/loggers/__init__.py]: Same.

File: train.py [hunk 3]
```diff
@@ -90,6 +90,10 @@ def train(hyp, opt, device, callbacks):  # hyp is path/to/hyp.yaml or hyp dictio
     data_dict = None
     if RANK in {-1, 0}:
         loggers = Loggers(save_dir, weights, opt, hyp, LOGGER)  # loggers instance
+        if loggers.clearml:
+            # Data_dict is either None is user did not choose for ClearML dataset or is filled in by ClearML
+            data_dict = loggers.clearml.data_dict
+
```
Comment (AyushExel) [inline train.py]: @thepycoder I didn't understand this part. If `loggers.clearml.data_dict`is None when users don't use clearml dataset management, then it'd make data_dict None which would break other parts right?

File: utils/general.py [hunk 4]
```diff
@@ -436,6 +436,9 @@ def check_file(file, suffix=''):
             torch.hub.download_url_to_file(url, file)
             assert Path(file).exists() and Path(file).stat().st_size > 0, f'File download failed: {url}'  # check
         return file
+    elif file.startswith('clearml://'):
+        # ClearML Dataset ID
+        return file
     else:  # search
```
Comment (AyushExel) [inline utils/general.py:456]: We should probably prepare for the edge case where user passes clearml file but clearml is not installed.

File: utils/loggers/wandb/wandb_utils.py [hunk 5]
```diff
@@ -43,6 +43,9 @@ def check_wandb_config_file(data_config_file):
 def check_wandb_dataset(data_file):
     is_trainset_wandb_artifact = False
     is_valset_wandb_artifact = False
+    if isinstance(data_file, dict):
+        # In that case another dataset manager has already processed it and we don't have to
+        return None
```
Comment (AyushExel) [inline utils/loggers/wandb/wandb_utils.py]: it should return the data_file instead of None right?

**Comment by AyushExel** [comment]
@thepycoder Hey I've taken a quick look. I haven't used clearML before so some of my comments might just be wrong.
Also, about comments, we'd like to keep them to bare minimum and only where needed in the core files. You can add more descriptive comments in the clearml_utils.py if needed. So, I've marked some places which would do fine without extra comments.

Also, can you share a colab that we can use to try and test the cases for this integration? Things like vanilla logging, artifacts, logging dataset, using clearml datasets

File: train.py (hunk 3 above)
Comment (thepycoder) [reply train.py]: You're right it would make `data_dict` None, but that's the goal. A little more down (L110), the original code checks if `data_dict` is set and if not, uses `check_dataset` to fill it. The None is basically just saying: it was not filled by any third party, so fallback to standard yolo dataset

```
    with torch_distributed_zero_first(LOCAL_RANK):
        data_dict = data_dict or check_dataset(data)  # check if None 
```

File: utils/general.py (hunk 4 above)
Comment (thepycoder) [reply utils/general.py:456]: You're 100% correct, thanks for catching it!

File: utils/loggers/wandb/wandb_utils.py (hunk 5 above)
Comment (thepycoder) [reply utils/loggers/wandb/wandb_utils.py]: Yup, I think you're right. In this case, it would be None whenever both ClearML and Wandb are installed and a `clearml://` url is given, which would indeed crash. I will have to test this though.

**Comment by thepycoder** [comment]
@AyushExel Thanks for the notes! I have address all of them and created a small notebook to test the functionality here: https://colab.research.google.com/drive/1ZNqeFVn8-MrsYCJLxmZ2q7Vy4uH3mHQV

File: train.py (hunk 3 above)
Comment (AyushExel) [reply train.py]: Thanks!

File: utils/loggers/__init__.py (hunk 2 above)
Comment (AyushExel) [reply utils/loggers/__init__.py]: @thepycoder Can you address this as well? You can move these comments to clearml_utils if you like

**Comment by AyushExel** [comment]
Thankyou! Looks good. I'll try it out now.
@glenn-jocher Do you want to take a look ?

**Comment by AyushExel** [comment]
@thepycoder running you notebook. I ran the cells in this order 1->2->4 and I get this
Screenshot 2022-08-01 at 7 15 40 PM

**Comment by thepycoder** [comment]
@AyushExel I'm sorry about that, fixed it in the last commit, but forgot the remove the import statement. Should be fixed now!

**Comment by AyushExel** [comment]
@thepycoder thanks, I tried your notebook and it works now. I have a question:
* Do I need to set up the API key every time I run the colab? I'm not sure if I'm doing something wrong, but I needed to not only create a new API config but also had to press enter 3-4 times to complete the setup as there are different inputs.

**Comment by thepycoder** [comment]
@AyushExel The multiple inputs thing is true, its just how the clearml-init command works on our end :)

For the API key: it is saved in a local file, so every time that file is lost (e.g. New colab) would mean a new api key. I haven't checked, is this not the same as with wandb? You have to supply a key again every time the machine is wiped, no?

**Comment by AyushExel** [comment]
@thepycoder not always.. I don't know the implementation details but I think wandb read .netrc file for credentials so you don't usually have to login every time you restart the colab. 
But that aside, I think the multiple inputs thing might confuse users who are already used to current logging. But overall the colab worked fine so I think this should be good to go once Glenn gets a chance to try this.

**Comment by thepycoder** [comment]
@AyushExel If you have to resupply an api key when the colab simply restarts (without wiping disk) then something is wrong. I'll check this in the morning, thanks for bringing it up!

Clearml-init is a command-line tool, so it indeed might look clunky through colab. I'll check with the devs if there's an option to add a flag that will skip the 3xenter presses. Should indeed make the process nicer :)

**Comment by AyushExel** [comment]
Oh one more thing, I think the readme and tutorial changes need to be discussed separately. Usually Glenn does these after the integration is live.

**Comment by glenn-jocher** [comment]
@thepycoder @AyushExel I'll see if I can resolve the tutorial conflicts

**Comment by glenn-jocher** [comment]
@thepycoder ok conflicts resolved and PR is updated to master. I'll review with @AyushExel tomorrow Tuesday and let you know if we need other changes

**Comment by glenn-jocher** [comment]
@thepycoder the 'Advanced' section in the Colab notebook is a broken/TODO link. Should we remove this section?
Screen Shot 2022-08-02 at 1 29 59 AM

**Comment by thepycoder** [comment]
@glenn-jocher The TODO url is actually for the github issue, but I didn't know when we were going to create it (similar to wandb).

Should we make a placeholder issue already, update the link and then finalize in a separate PR?

@AyushExel 
>Oh one more thing, I think the readme and tutorial changes need to be discussed separately. Usually Glenn does these after the integration is live.

Didn't know this, let me know if there are parts that need to be cut out of the PR

**Comment by glenn-jocher** [comment]
@AyushExel ok I've simplified the 'W&B Logging' and 'Local Logging' areas of the Colab notebook in master and merged master into this PR. They're more vertically concise now and in line with our new policy of 1 full width image at around 16:9 or 4:3 aspect ratio. Do you think we should eliminate the results.png visual from the notebook also?

@thepycoder please take a look at the updated notebook and select only one image to display in the ClearML Logging section (just for tutorial.ipynb, your ClearML markdown file can have as much content as you want). The new images in the notebook are 1920 pixel-wide JPGs at about 500kb, so if you can shoot for that perfect.

About your other question the 'Issue' we usually open is just a copy of the README/Markdown/Wiki you already have. This is a bit unconventional, the idea is to simply allows users to comment/question directly on the topic. Let me think about this a bit as we may want to move away from this policy as it duplicates maintenance and would leave ClearML content at risk of diverging over time if not adequately maintained in both places. 

A key point here is that we don't need this PR to be a 100% complete solution, we should ensure it is as bug free as possible with adequate documentation, and from there I'm sure we will continue to improve off this baseline with future fixes/features.

**Comment by thepycoder** [comment]
@glenn-jocher I've restructured the ClearML section, there is now only 1 image, a 1920 wide JPG of 431kb, ought to be close enough right? :smile: 

For the links to the (non-existent) issue, I have corrected them to: https://docs.ultralytics.com/yolov5/tutorials/clearml_logging_integration which should be the correct link to our tutorial README once its merged in master. So naturally this link will not work yet as long as the PR isn't merged. Is this OK?

I 100% agree, if you are good with the state as it is now, feel free to merge it. If anything else comes up, I can always add fixes, just ping me! :)

**Comment by AyushExel** [comment]
@glenn-jocher yes the results.png is not needed there. Maybe just add a list of all visualization files generated in the "local logging" section without displaying all of them

**Comment by thepycoder** [comment]
@glenn-jocher Is there anything you need me to do still?

**Comment by glenn-jocher** [comment]
@thepycoder let's set up a quick meeting to finalize everything. Can you select a time here? Thanks!
https://ultralytics.com/meet

**Comment by glenn-jocher** [comment]
@thepycoder I've reviewed and updated the PR as best I could. I'll merge now, but please review and verify I haven't broken any functionality, and please work on updates to streamline ClearML ops as mentioned over email so we can improve the user experience for ClearML logging.

FYI inserted links for improved analytics for us:
Screen Shot 2022-08-05 at 8 38 36 PM

**Comment by glenn-jocher** [comment]
@thepycoder I've removed the plot titles from labels.png and labels_correlogram.png as they were applied on the last subplot only. All other plot titles are nice additions, thanks!

labels_correlogram
labels

**Comment by glenn-jocher** [comment]
@thepycoder PR is merged. Thank you for your contributions to YOLOv5 🚀 and Vision AI ⭐

**Comment by glenn-jocher** [comment]
@thepycoder I noticed the light/dark mode graphics look good in GitHub (very clever), but unfortunately do not extend well to other platforms with automatic README views that we use like Docker Hub and Paperspace Gradient (see links below). Can you please provide a single graphic for use here? Maybe the light mode graphic with an opaque white inside the circle would work.

- https://console.paperspace.com/github/ultralytics/yolov5
- https://hub.docker.com/repository/docker/ultralytics/yolov5

**Comment by Robotatron** [comment]
Will ClearML be supported for segmentation as well? If so, any ETA on this? @thepycoder @glenn-jocher

**Comment by thepycoder** [comment]
Hey @Robotatron! What features would you like to see specifically? I can def take a look :)

**Comment by Robotatron** [comment]
> Hey @Robotatron! What features would you like to see specifically? I can def take a look :)

Hey @thepycoder, thanks for showing interest. The features would be all the common stuff you can think of (probably the same stuff you log with object detection for ClearML already), in order of importance:

1) Current YOLO5 segmentation Tensorboard scalar metrics that can be visualized in a graph with ClearML -> so losses and segmentation mAPs (AP50:95, AP50, etc). The ones YOLO5 logs into "`results.csv`" file 
2) opt.yaml, the config of the run
3) If possible, per class mAP metrics (we get those printed in the console when running inference, but not during training sadly)
4) Inference images on the test set (the one we get with YOLO automatically in the experiment folder)
5) Train batch images (the one we get with YOLO automatically in the experiment folder)
6) (this will be probably too much, but prediction masks on the test set in the COCO format)

**Comment by thepycoder** [comment]
@Robotatron , please check out #10752 and let me know if it works for you! :)

**Comment by Robotatron** [comment]
@thepycoder Everything works great, thank you! 
Would you know if I have to configure clearML or edit your logger to also log the best model under "artefacts"? I am new to ClearML so could be a stupid question :)

**Comment by thepycoder** [comment]
@Robotatron The best model should always be logged under the artifacts tab. If not, there's a bug somewhere.
To get the latest model too you'll have to set a save inteval using the yolo arguments themselves :)
